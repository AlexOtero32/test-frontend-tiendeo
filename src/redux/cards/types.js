export const CREATE_CARD = 'create_card';
export const EDIT_CARD = 'edit_card';
export const DELETE_CARD = 'delete_card';
export const SET_EDITING_CARD = 'set_editing_card';
export const CLEAR_EDITING_CARD = 'clear_editing_card';
