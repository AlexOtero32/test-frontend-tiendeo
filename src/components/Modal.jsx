import React from 'react';

const Modal = ({ onClose }) => {
    return (
        <form className="bg-white shadow-md rounded px-8 pt-6 pb-8 mb-4">
            <div className="form-group">
                <label className="label">Título</label>
                <input
                    className="input"
                    id="titulo"
                    type="text"
                    placeholder="Título"
                    required
                />
            </div>
            <div className="form-group">
                <label className="label">Descripcion</label>
                <textarea
                    className="input"
                    id="descripcion"
                    placeholder="Descripción"
                    required
                />
            </div>
            <div className="form-group">
                <label className="label">URL de la imagen</label>
                <input
                    className="input"
                    id="descripcion"
                    placeholder="URL"
                    type="text"
                    required
                />
            </div>
            <p className="text-red-500 text-xs italic">
                Título y descripción obligatorios.
            </p>
            <div className="flex items-center justify-between">
                <button
                    className="inline-block align-baseline font-bold text-sm text-blue-500 hover:text-blue-800"
                    onClick={() => onClose(false)}
                >
                    Cancelar
                </button>
                <button className="button button-blue" type="button">
                    Crear
                </button>
            </div>
        </form>
    );
};

export default Modal;
