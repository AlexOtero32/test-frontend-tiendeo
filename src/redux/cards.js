import { v4 as uuid } from 'uuid';

export const CREATE_CARD = 'create_card';
export const EDIT_CARD = 'edit_card';
export const DELETE_CARD = 'delete_card';

const initialState = {
    cards: [1, 2, 3],
};

export default (state = initialState, action) => {
    switch (action.type) {
        case CREATE_CARD:
            const card = {
                ...action.card,
                id: uuid(),
                fecha: new Date(),
            };

            state.cards.push(card);

            return state;
        case EDIT_CARD:
            return state;
        case DELETE_CARD:
            return state;
        default:
            return state;
    }
};
