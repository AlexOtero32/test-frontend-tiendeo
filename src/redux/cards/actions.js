/**
 * Action creators
 */

import {
    CREATE_CARD,
    EDIT_CARD,
    DELETE_CARD,
    SET_EDITING_CARD,
    CLEAR_EDITING_CARD,
    ORDER_CARDS,
} from './types';

export function createCard(card) {
    return { type: CREATE_CARD, card };
}

export function editCard(card) {
    return { type: EDIT_CARD, card };
}

export function deleteCard(cardId) {
    return { type: DELETE_CARD, cardId };
}

export function setEditingCard(card) {
    return { type: SET_EDITING_CARD, card };
}

export function clearEditingCard() {
    return { type: CLEAR_EDITING_CARD };
}

export function orderBy(order) {
    return { type: ORDER_CARDS, order };
}
