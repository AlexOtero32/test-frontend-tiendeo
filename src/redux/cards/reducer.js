import { v4 as uuid } from 'uuid';
import {
    CREATE_CARD,
    EDIT_CARD,
    DELETE_CARD,
    SET_EDITING_CARD,
    CLEAR_EDITING_CARD,
} from './types';

const initialState = {
    cardList: [],
    editingCard: {},
};

export default (state = initialState, action) => {
    switch (action.type) {
        case CREATE_CARD:
            const card = {
                ...action.card,
                id: uuid(),
                fecha: new Date(),
            };

            return {
                cardList: [...state.cardList, card],
                editingCard: state.editingCard,
            };

        case EDIT_CARD:
            const cardList = state.cardList.map(function (card) {
                if (card.id === action.card.id) {
                    Object.assign(card, action.card);
                }
                return card;
            });

            return {
                cardList,
                editingCard: {},
            };

        case DELETE_CARD:
            return {
                cardList: state.cardList.filter(
                    (card) => card.id !== action.cardId
                ),
                editingCard: state.editingCard,
            };

        case SET_EDITING_CARD:
            return { cardList: state.cardList, editingCard: action.card };

        case CLEAR_EDITING_CARD:
            return { cardList: state.cardList, editingCard: {} };

        default:
            return state;
    }
};
