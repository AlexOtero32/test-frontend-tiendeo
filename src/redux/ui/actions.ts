import { ActionTypes, Themes } from "./constants";

interface ToggleModal {
  type: typeof ActionTypes.TOGGLE_MODAL;
}
interface SwitchTheme {
  type: typeof ActionTypes.SWITCH_THEME;
  theme: Themes;
}

export type UIActions = ToggleModal | SwitchTheme;
