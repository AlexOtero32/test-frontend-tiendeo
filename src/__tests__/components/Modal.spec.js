import React from "react";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";

import Modal from "../../components/Modal";
import { clearEditingCard } from "../../redux/cards/actions";

const mockStore = configureStore([]);

describe("Modal component", function () {
  let store;

  const initialState = {
    ui: {
      isModalOpen: false,
    },
    cards: {
      orderBy: "newest",
      editingCard: {},
      cardList: [
        {
          date: new Date(),
          titulo: "Tarjeta",
          id: "un-id-aleatorio",
          descripcion: "DescripciÃ³n",
          url: undefined,
        },
      ],
    },
  };

  beforeEach(() => {
    store = mockStore(initialState);
  });

  it("Renderiza correctamente", function () {
    const component = renderer.create(
      <Provider store={store}>
        <Modal />
      </Provider>
    );

    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("Limpia editingCard al pulsar en cancelar", function () {
    store.dispatch = jest.fn();

    const component = renderer.create(
      <Provider store={store}>
        <Modal />
      </Provider>
    );

    renderer.act(() => {
      component.root.findAllByType("button")[0].props.onClick();
    });

    expect(store.dispatch).toHaveBeenLastCalledWith(clearEditingCard());
  });
});
