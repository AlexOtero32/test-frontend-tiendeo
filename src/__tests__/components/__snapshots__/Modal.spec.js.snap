// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component Renderiza correctamente 1`] = `
<form
  className="form"
  method="POST"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      className="label"
    >
      Título
    </label>
    <input
      className="input"
      name="titulo"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="Título"
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="label"
    >
      Descripción
    </label>
    <textarea
      className="input"
      name="descripcion"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="Descripción"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="label"
    >
      URL
    </label>
    <input
      className="input"
      name="url"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="https://ejemplo.com/imagen.png"
      type="text"
      value=""
    />
  </div>
  <div
    className="flex items-center justify-between"
  >
    <button
      className="inline-block align-baseline font-bold text-sm text-blue-500 hover:text-blue-800"
      onClick={[Function]}
    >
      Cancelar
    </button>
    <button
      className="button button-blue"
      type="submit"
    >
      Guardar
    </button>
  </div>
</form>
`;
